cpp/src/io/csv/csv_gpu.cu

```diff
@@ -570,6 +571,329 @@ __global__ void convertCsvToGdf(const char *raw_csv,
+ **/
+inline __device__ uint32_t make_short_row_context(uint32_t id0,
+                                                  uint32_t id1,
+                                                  uint32_t id2 = ROW_CTX_COMMENT,
+                                                  uint32_t c0  = 0,
+                                                  uint32_t c1  = 0,
+                                                  uint32_t c2  = 0)
+{
+  return (id0 << 8) | (id1 << 10) | (id2 << 12) | (ROW_CTX_EOF << 14) | (c0) | (c1 << 1) |
+         (c2 << 2);
+}
+
+/**
+ * @brief Merge a short row context to keep track of bitmasks where new rows occur
+ * Merges a single-character "block" row context at position pos with the current
+ * block's row context (the current block contains 32-pos characters)
+ *
+ * @param ctx Current block context and new rows bitmaps
+ * @param ctx_short state transitions associated with new character
+ * @param pos Position within the current 32-character block
+ *
+ * NOTE: This is probably the most performance-critical piece of the row gathering kernel.
+ * The ctx_short value should be created via make_short_row_context, and its value should
+ * have been evaluated at compile-time.
+ *
+ **/
+inline __device__ void merge_short_row_context(uint4 &ctx, uint32_t ctx_short, uint32_t pos)
+{
+  uint32_t id0 = (ctx.w >> 0) & 3;
+  uint32_t id1 = (ctx.w >> 2) & 3;
+  uint32_t id2 = (ctx.w >> 4) & 3;
+  ctx.x |= ((ctx_short >> id0) & 1) << pos;
+  ctx.y |= ((ctx_short >> id1) & 1) << pos;
+  ctx.z |= ((ctx_short >> id2) & 1) << pos;
+  ctx.w = ((ctx_short >> (8 + id0 * 2)) & 0x03) | ((ctx_short >> (6 + id1 * 2)) & 0x0c) |
+          ((ctx_short >> (4 + id2 * 2)) & 0x30) | (ROW_CTX_EOF << 6);
+}
+
+/**
+ * @brief 512-wide row context merge transform
+ *
+ * Repeatingly merge row context blocks, keeping track of each merge operation
+ * in a context tree so that the transform is reversible
+ * The tree is organized such that the left and right children of node n
+ * are located at indices n*2 and n*2+1, the root node starting at index 1
+ *
+ * Each node contains the counts and output contexts corresponding to the
+ * possible input contexts.
+ * Each parent node's count is obtained by adding the corresponding counts
+ * from the left child node with the right child node's count selected from
+ * the left child node's output context:
+ *   parent.count[k] = left.count[k] + right.count[left.outctx[k]]
+ *   parent.outctx[k] = right.outctx[left.outctx[k]]
+ *
+ * @param ctxtree[out] packed row context tree
+ * @param ctxb[in] packed row context for the current character block
+ * @param t thread id (leaf node id)
+ *
+ **/
+static inline __device__ void rowctx_merge_transform(uint64_t ctxtree[1024],
+                                                     packed_rowctx_t ctxb,
+                                                     uint32_t t)
+{
+  uint64_t tmp;
+
+#define CTX_MERGE(lanemask, tmask, base, level_scale)                       \
```

Comment:
I think this could be a lambda function and still generate the same/similar inlined code. Not requesting a change, just a thinking out loud.

Reply:
It'd be ok with me, but I'd want to see proof at godbolt that it does :) (I tend to not trust the compiler by default for this sort of things) ~ And also that it doesn't double the lines of code, otherwise I don't see the point.

Reply:
Right, just thinking out loud, not actually requesting a change. I know some people prefer to avoid macros when at all possible, so I mentioned it.

Reply:
I know some people prefer to avoid lamdbas when at all possible [*insert "I know him, he is me" meme] :)

Reply:
Even a `constexpr inline` function would obviously be inlined -- the inputs are all literals.

Reply:
Could also be a macro. It seems to me that this is the solution that results in the least amount of code. Less code = less bugs.

Reply:
What is the additional code that an inline function would have over the macro? I can only think of the parameter types, which is a positive IMO.